lib/icinga/clusterevents-check.cpp

```diff
@@ -171,25 +272,74 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
 		try {
 			host->ExecuteRemoteCheck(macros);
 		} catch (const std::exception& ex) {
-			CheckResult::Ptr cr = new CheckResult();
-			cr->SetState(ServiceUnknown);
-
 			String output = "Exception occurred while checking '" + host->GetName() + "': " + DiagnosticInformation(ex);
-			cr->SetOutput(output);
-
+			ServiceState state = ServiceUnknown;
 			double now = Utility::GetTime();
-			cr->SetScheduleStart(now);
-			cr->SetScheduleEnd(now);
-			cr->SetExecutionStart(now);
-			cr->SetExecutionEnd(now);
 
-			Dictionary::Ptr message = MakeCheckResultMessage(host, cr);
-			listener->SyncSendMessage(sourceEndpoint, message);
+			if (params->Contains("source")) {
+				SendEventExecuteCommand(params, state, output, now, now, listener, origin, sourceEndpoint);
+			} else {
+				CheckResult::Ptr cr = new CheckResult();
+				cr->SetState(state);
+				cr->SetOutput(output);
+				cr->SetScheduleStart(now);
+				cr->SetScheduleEnd(now);
+				cr->SetExecutionStart(now);
+				cr->SetExecutionEnd(now);
+
+				Dictionary::Ptr message = MakeCheckResultMessage(host, cr);
+				listener->SyncSendMessage(sourceEndpoint, message);
+			}
 
 			Log(LogCritical, "checker", output);
 		}
 	} else if (command_type == "event_command") {
-		host->ExecuteEventHandler(macros, true);
+		try {
+			host->ExecuteEventHandler(macros, true);
+		} catch (const std::exception& ex) {
+			if (params->Contains("source")) {
+				String output = "Exception occurred while executing event command '" + command + "' for '" +
+					host->GetName() + "': " + DiagnosticInformation(ex);
+
+				double now = Utility::GetTime();
+				SendEventExecuteCommand(params, ServiceUnknown, output, now, now, listener, origin, sourceEndpoint);
+			} else {
+				throw;
```

Comment:
This will just kill Icinga without logging anything. Please at least throw a helpful exception.

Reply:
`git diff --ignore-all-space 78a94219aa4470833b89cef784eacc59c869f778..HEAD -- lib/icinga/clusterevents-check.cpp` clearly shows...

```
        } else if (command_type == "event_command") {
+               try {
                        host->ExecuteEventHandler(macros, true);
+               } catch (const std::exception& ex) {
+                       if (params->Contains("source")) {
+                               String output = "Exception occurred while executing event command '" + command + "' for '" +
+                                       host->GetName() + "': " + DiagnosticInformation(ex);
+
+                               double now = Utility::GetTime();
+                               SendEventExecuteCommand(params, ServiceUnknown, output, now, now, listener, origin, sourceEndpoint);
+                       } else {
+                               throw;
+                       }
+               }
```

... that the whole try-catch itself, not just that `throw;` has been added by this PR. So the `throw;` is the old behavior.

I consider this another construction area.